lib/Backend/arm64/EncoderMD.cpp

```diff
@@ -321,10 +321,58 @@ int EncoderMD::EmitOp3RegisterOrImmediate(Arm64CodeEmitter &Emitter, IR::Instr*
     }
 }
 
+template<typename _RegFunc32, typename _RegFunc64, typename _ImmFunc32, typename _ImmFunc64>
+int EncoderMD::EmitOp3RegisterOrImmediateExtendSPReg(Arm64CodeEmitter &Emitter, IR::Instr* instr, _RegFunc32 reg32, _RegFunc64 reg64, _ImmFunc32 imm32, _ImmFunc64 imm64)
+{
+    // if we have two regopnds as sources, then we need to be careful
+    if (instr->GetSrc1()->IsRegOpnd() && instr->GetSrc2()->IsRegOpnd())
+    {
+        IR::RegOpnd* src1 = instr->GetSrc1()->AsRegOpnd();
+        IR::RegOpnd* src2 = instr->GetSrc2()->AsRegOpnd();
+        IR::RegOpnd* dst  = instr->GetDst()->AsRegOpnd();
+        AssertMsg(!(src1->GetReg() == RegSP && src2->GetReg() == RegSP), "Tried to encode an add or sub that used RegSP as both sources - tighten legalization restrictions!");
+
+        // We need to swap the parameters if
+        // 1. src2 is RegSP
+        // 2. src1 is RegZR and dst is RegSP
+        // This is because the valid instruction forms are
+        // add Rd, Rs, Rw_SFT
+        // add Rd|SP, Rs|SP, Rw_EXT
+        // and the encoding for SP is the same as the encoding for ZR
+        if (src2->GetReg() == RegSP || (src1->GetReg() == RegZR && dst->GetReg() == RegSP))
```

Comment:
I think this swapping logic would fit better in the legalizer, with asserts in the encoder to verify it was done.

Reply:
Generally yes, but we'd still need to do a bunch of stuff here, as well as complicate the legalization code since the semantics of which registers affect which registers can be placed elsewhere is not really present there yet.